boogie_ast/src/boogie_program/mod.rs

```diff
@@ -0,0 +1,279 @@
+/// Boogie types
+pub enum Type {
+    /// Boolean
+    Bool,
+
+    /// Bit-vector of a given width, e.g. `bv32`
+    Bv(usize),
+
+    /// Unbounded integer
+    Int,
+
+    /// Map type, e.g. `[int]bool`
+    Map { key: Box<Type>, value: Box<Type> },
+}
+
+/// Function and procedure parameters
+pub struct Parameter {
+    name: String,
+    typ: Type,
+}
+
+/// Literal types
+pub enum Literal {
+    /// Boolean values: `true`/`false`
+    Bool(bool),
+
+    /// Bit-vector values, e.g. `5bv8`
+    Bv {
+        width: usize,
+        value: String, // TODO: use bigint
+    },
+
+    /// Unbounded integer values, e.g. `1000`
+    Int(String), // TODO: use bigint
+}
+
+/// Unary operators
+pub enum UnaryOp {
+    /// Logical negation
+    Not,
+
+    /// Arithmetic negative
+    Neg,
+}
+
+pub enum BinaryOp {
+    /// Logical AND
+    And,
+
+    /// Logical OR
+    Or,
+
+    /// Equality
+    Eq,
+
+    /// Inequality
+    Neq,
+
+    /// Less than
+    Lt,
+
+    /// Less than or equal
+    Lte,
+
+    /// Greater than
+    Gt,
+
+    /// Greater than or equal
+    Gte,
+
+    /// Addition
+    Add,
+
+    /// Subtraction
+    Sub,
+
+    /// Multiplication
+    Mul,
+
+    /// Division
+    Div,
+
+    /// Modulo
+    Mod,
+}
+
+/// Expr types
+pub enum Expr {
+    /// Literal (constant)
+    Literal(Literal),
+
+    /// Variable
+    Symbol { name: String },
+
+    /// Unary operation
+    UnaryOp { op: UnaryOp, operand: Box<Expr> },
+
+    /// Binary operation
+    BinaryOp { op: BinaryOp, left: Box<Expr>, right: Box<Expr> },
+}
+
+/// Statement types
+pub enum Stmt {
+    /// Assignment statement: `target := value;`
+    Assignment { target: String, value: Expr },
+
+    /// Assert statement: `assert condition;`
+    Assert { condition: Expr },
+
+    /// Assume statement: `assume condition;`
+    Assume { condition: Expr },
+
+    /// Statement block: `{ statements }`
+    Block { statements: Vec<Stmt> },
+
+    /// Break statement: `break;`
+    /// A `break` in boogie can take a label, but this is probably not needed
+    Break,
+
+    /// Procedure call: `symbol(arguments);`
+    Call { symbol: String, arguments: Vec<Expr> },
+
+    /// Declaration statement: `var name: type;`
+    Decl { name: String, typ: Type },
+
+    /// If statement: `if (condition) { body } else { else_body }`
+    If { condition: Expr, body: Box<Stmt>, else_body: Option<Box<Stmt>> },
+
+    /// Goto statement: `goto label;`
+    Goto { label: String },
+
+    /// Label statement: `label:`
+    Label { label: String },
+
+    /// Return statement: `return;`
+    Return,
+
+    /// While statement: `while (condition) { body }`
+    While { condition: Expr, body: Box<Stmt> },
+}
+
+/// Procedure specification
+pub struct Contract {
+    /// Pre-conditions
+    requires: Vec<Expr>,
+    /// Post-conditions
+    ensures: Vec<Expr>,
+    /// Modifies clauses
+    // TODO: should be symbols and not expressions
+    modifies: Vec<Expr>,
+}
+
+/// Procedure definition
```

Comment:
Can you please document what a procedure is in Boogie?

Reply:
Done

Reply:
Thanks!